engine/access/state_stream/backend/backend_executiondata2_test.go

```diff
@@ -0,0 +1,1065 @@
+
+		block := s.blocksHeightToBlockMap[currentHeight]
+		expectedExecutionData := s.blockIDToExecutionDataMap[block.ID()]
+		require.Equal(s.T(), expectedExecutionData, actualExecutionData.ExecutionData)
+
+		currentHeight += 1
+
+		received++
+		if received == expected {
+			// We’ve validated enough; stop the stream.
+			cancel()
+		}
+	}
+
+	require.ErrorIs(s.T(), sub.Err(), context.Canceled)
+}
+
+// TestGetExecutionData verifies the retrieval of execution data for a specific block ID.
+// It checks that the data and metadata are returned correctly.
+func (s *BackendExecutionDataSuite2) TestGetExecutionData() {
+	s.mockDataProviderState()
+
+	backend := s.createExecutionDataBackend()
+
+	actualExecData, metadata, err :=
+		backend.GetExecutionDataByBlockID(context.Background(), s.sporkRootBlock.ID(), s.criteria)
+	expectedExecData := s.blockIDToExecutionDataMap[s.sporkRootBlock.ID()]
+
+	require.Nil(s.T(), err)
+	require.NotEmpty(s.T(), metadata)
+	require.Equal(s.T(), expectedExecData, actualExecData)
+}
+
+// TestGetExecutionData_Errors verifies how GetExecutionDataByBlockID handles various error
+// conditions, such as missing data, block not found, fork abandoned, and unexpected system errors.
+func (s *BackendExecutionDataSuite2) TestGetExecutionData_Errors() {
+	backend := s.createExecutionDataBackend()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	block := s.sporkRootBlock
+
+	s.Run("execution result info returns data not found", func() {
+		// provider returns storage.ErrNotFound which should map to access.DataNotFoundError
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(nil, storage.ErrNotFound).
+			Once()
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctx, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		require.Error(s.T(), err)
+		require.True(s.T(), access.IsDataNotFoundError(err))
+	})
+
+	s.Run("execution result info returns block not found", func() {
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(nil, optimistic_sync.ErrBlockNotFound).
+			Once()
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctx, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		require.Error(s.T(), err)
+		require.True(s.T(), access.IsDataNotFoundError(err))
+	})
+
+	s.Run("execution result info returns fork abandoned", func() {
+		s.executionResultProvider.
+			On("ExecutionResultInfo", block.ID(), mock.Anything).
+			Return(nil, optimistic_sync.ErrForkAbandoned).
+			Once()
+
+		execDataRes, metadata, err :=
+			backend.GetExecutionDataByBlockID(ctx, block.ID(), s.criteria)
+		assert.Nil(s.T(), execDataRes)
+		assert.Nil(s.T(), metadata)
+		require.Error(s.T(), err)
+		require.True(s.T(), access.IsPreconditionFailedError(err))
```

Comment:
I don't think `PreconditionFailed` is the right error for an abandoned fork. This generally means that the current state does not support the action, but it may in the future, whereas an abandoned fork indicates the data will never be valid.